fix: resolve data race in Broadcasters system

cwaldren-ld opened the pull request:
This resolves a data race in the Broadcaster implementation, specifically ` HasListeners()` was not protected by a lock. 

Additionally it swaps out the mutex for a `RWLock` so that concurrent readers don't block each other. 

A behavioral change is that the `Broadcast` method now locks the subscribers array, so that `Add` or `RemoveListener` cannot be called concurrently.

JohnStarich committed:
Fix broadcast data races (#151)

**Requirements**

- [x] I have added test coverage for new or changed functionality
- [x] I have followed the repository's [pull request submission
guidelines](../blob/v5/CONTRIBUTING.md#submitting-pull-requests)
- [ ] I have validated my changes against all supported platform
versions
* What does "platform" mean here? Go version? OS version? Happy to help
test, just not sure what I'm testing!

**Related issues**

I hit data races when running tests in parallel, which effectively
obscures other data races my application may have. It looks like others
did too, judging by #102.
Fixes https://github.com/launchdarkly/go-server-sdk/issues/102

**Describe the solution you've provided**

I've addressed 2 data races with improved locking:
* `HasListeners()` had a data race due to any mutation on
`b.subscribers`
* `Close()` had a data race where closing a send channel triggers a
panic in `Broadcast()`

I also saw an easy opportunity to use more fine-grained locks with an
RWMutex, although I'm happy to back that out if you would prefer.

**Describe alternatives you've considered**

I also considered using an atomic data type for subscribers, but I
figured that change would be less of a surgical fix. It also may be more
difficult to mutate the slice, since a compare-and-swap can fail and
would need a loop (it's not as simple as `atomic.Add`).

Another idea which may be simpler is using a channel to manage shutdown.
Typically I'd use a `context.Context` here to manage cancellation.
That'd also prevent `Broadcast()`ing on a full send channel from
blocking `Close()`.

**Additional context**

Add any other context about the pull request here.

cwaldren-ld committed:
Additional test for Broadcaster data race (#152)

This is a small cleanup to the existing Broadcasters implementation to
use `slices.DeleteFunc`.

cwaldren-ld committed:
ensure receiver channel is drained

kinyoklion commented on internal/broadcasters.go:
```diff
@@ -50,35 +49,31 @@ func (b *Broadcaster[V]) AddListener() <-chan V {
-		for _, ch := range ss {
-			ch.sendCh <- value
-		}
+	b.lock.RLock()
```
So, there is a behavioral change here, which may or may not be relevant. But previously during the broadcast you could remove a listener (the lock wasn't retained during the actual broadcast), and now you wouldn't be able do get the exclusive lock during that time.

cwaldren-ld replied:
hmm, we probably shouldn't hold the lock during the broadcast as that could take an unbounded amount of time to complete. I'll revert that and add a test.

cwaldren-ld replied:
Ok, so there's a problem here: if we allow removing a listener while Broadcast is running concurrently, then it could try to send to a channel that is closed (due to removing the listener.) This is a panic. 

So that behavior was already bad. 

I'm thinking it's better to say "if you aren't draining your receivers, you're not gonna be able to add or remove receivers" which would be the case if we locked around the broadcast.

cwaldren-ld replied:
I'm reverting to locking around the broadcast. I'll note in the release that this may affect the use-case of adding/removing a listener while broadcasting.

cwaldren-ld committed:
Remove random order test

This test is fundamentally flawed in that the order of executing the
broadcasters is important. We can't execute them in random order without
causing deadlock, as receivers need to drain their channels for
broadcast to unblock.

cwaldren-ld committed:
add test for concurrent AddListener and Broadcast

cwaldren-ld committed:
revert to locking around the broadcast
